File: Titanic Script.R
# Load the dataset and the SDSFoundations package. 
# We added stringsAsFactors=FALSE so as not to create a NULL category in Embarked variable
dataset_train <- read.csv("data/train.csv", stringsAsFactors=FALSE)
library("SDSFoundations", lib.loc="C:/Program Files/R/R-3.4.3/library")

# View the column names of the dataset
names(dataset_train)

# View the Pclass distribution
table(dataset_train$Pclass)
round(prop.table(table(dataset_train$Pclass)), 3)
barplot(table(dataset_train$Pclass), main = "Passenger Class Distribution", xlab = "Class", ylab = "Count")

# View the Survived distribution
table(dataset_train$Survived)
tbl <- round(prop.table(table(dataset_train$Survived)), 3)
lbls <- paste(c("No", "Yes"), "\n", tbl * 100, "%", sep="")
pie(table(dataset_train$Survived), main = "Survival Distribution", labels = lbls)

# View the Sex distribution
table(dataset_train$Sex)
tbl <- round(prop.table(table(dataset_train$Sex)), 3)
lbls <- paste(c("Female", "Male"), "\n", tbl * 100, "%", sep="")
pie(table(dataset_train$Sex), main = "Sex Distribution", labels = lbls)

# View the Embarked distribution
table(dataset_train$Embarked)
tbl <- round(prop.table(table(dataset_train$Embarked)), 3)
lbls <- paste(c("No info", "Cherbourg", "Queenstown", "Southampton"), "\n", tbl * 100, "%", sep="")
pie(table(dataset_train$Embarked), main = "Embarked Distribution", labels = lbls)

# View the Age distribution
hist(dataset_train$Age, breaks = 20, main = "Histogram of Age", xlab = "Age")

# View the Fare distribution
hist(dataset_train$Fare, xlab = "Fare", ylab = "Count", main = "Histogram of Fare")

# Check for independence between variables sex and survived
tbl <- table(dataset_train$Sex, dataset_train$Survived)
chisq.test(tbl, correct = FALSE)$expected # Check that expected counts are over 5
chisq.test(tbl, correct = FALSE) # The variables are not independent...

# Convert variable Pclass from numerical to categorical
dataset_train$Pclass <- as.factor(dataset_train$Pclass)
levels(dataset_train$Pclass)

# Create a calculated field to use column Cabin in our model
dataset_train$HasCabin <- as.factor(ifelse (dataset_train$Cabin != "", "Yes", "No"))

# Check for independence between variables HasCabin and PClass
tbl <- table(dataset_train$HasCabin, dataset_train$Pclass)
chisq.test(tbl, correct = FALSE)$expected
chisq.test(tbl, correct = FALSE) # The variables are not independent.

# Subset the Dataset to exclude lines with no value for Age and Embarked
dataset_train_noNULLS <- dataset_train[complete.cases(dataset_train$Age) & dataset_train$Embarked != "",]
summary(dataset_train_noNULLS)

# Convert variable Embarked from string to categorical
dataset_train_noNULLS$Embarked <- as.factor(dataset_train_noNULLS$Embarked)
levels(dataset_train_noNULLS$Embarked)

# Create a logistic regression model to predict survival
mdl <- glm(Survived ~ Pclass + Sex + Age + SibSp + Parch + Fare + HasCabin + Embarked, 
           data = dataset_train_noNULLS, family = "binomial")
library(car)
vif(mdl) # Test for multicolinearity. As expected PClass and HasCabin are somewhat correlated.
summary(mdl)

# We use the backwards elimination procedure to refine our model:
# we eliminate the variable with the highest p > 0.05 and check AIC in every iteration.

# Our final model. 
mdl <- glm(Survived ~ Pclass + Sex + Age + SibSp, data = dataset_train_noNULLS, family = "binomial")
vif(mdl)
summary(mdl)

# Plot of predicted probabilities of survival vs actual survival
plot(fitted.values(mdl), dataset_train_noNULLS$Survived, xlab = 'Predicted chance of survival', ylab = 'Actual Survival', 
     main = 'Plot of predicted probabilities of survival vs actual survival')

# Create survived hat column in our dataset
dataset_train_noNULLS$prob_of_survival <- fitted.values(mdl)
dataset_train_noNULLS$survived_hat <- ifelse(fitted.values(mdl) > 0.5, 1, 0)

# Create confusion matrix
confusion_matrix <- table(dataset_train_noNULLS$survived_hat, dataset_train_noNULLS$Survived)
round(prop.table(confusion_matrix), 2)

# Write the new dataset to an excel file, to create our CAP curve.
# library(openxlsx)
# write.xlsx(dataset_train_noNULLS,file = "dataset_train.xlsx")

#####################################
# Test our model against dataset_test and

# Import the test datasets.
dataset_test <- read.csv("data/test.csv")
test_survived_actuals <- read.csv("data/test survived actuals.csv")

# Convert Pclass to categorical in the test dataset
dataset_test$Pclass <- as.factor(dataset_test$Pclass)

# Fill out the missing values of age with the mean
m <- mean(dataset_test$Age,na.rm = T)
dataset_test$Age <- ifelse(is.na(dataset_test$Age), m,dataset_test$Age)

# Use our model to predict survival probabilities.
dataset_test$p_hat <- predict.glm(mdl,dataset_test,type = "response")

#  Compare predicted to actual survival
dataset_test$survived <- test_survived_actuals$Survived
dataset_test$survived_hat <- ifelse(dataset_test$p_hat > 0.5, 1, 0)

# Get our confusion matrix
confusion_matrix <- table(dataset_test$survived_hat, dataset_test$survived)
round(prop.table(confusion_matrix), 2)

File: Titanic Script.py
# Load the datasets and the required packages.
from sklearn import metrics
from sklearn.feature_selection import RFE
from sklearn.linear_model import LogisticRegression
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats
dataset_train = pd.read_csv('data\\train.csv')
dataset_test = pd.read_csv('data\\test.csv')

# View the column names and their types
dataset_train.info()

# View the first five lines
dataset_train.head()

# Check missing values in train data
dataset_train.isnull().sum()

# Convert variable Pclass from numerical to categorical
dataset_train.Pclass = dataset_train.Pclass.astype('category')

# View the Pclass distribution
Pclass_plot = sns.countplot(
    x='Pclass', data=dataset_train, palette="Pastel1_r")

# View the Survived distribution per Sex
dataset_train['Survived'] = dataset_train['Survived'].replace(
    0, 'No')
dataset_train['Survived'] = dataset_train['Survived'].replace(
   1, 'Yes')
sns.countplot(x='Survived', hue='Sex', data=dataset_train)

# View the Sex distribution
survived_count = dataset_train['Sex'].value_counts()
labels = 'Male', 'Female'
plt.pie(x=survived_count, labels=labels, autopct='%1.1f%%',
        colors=['lightcoral', 'lightskyblue'])
plt.show()

# View the Embarked distribution. (We rename the values of the variable).
dataset_train['Embarked'] = dataset_train['Embarked'].replace('C', 'Cherbourg')
dataset_train['Embarked'] = dataset_train['Embarked'].replace(
    'Q', 'Queenstown')
dataset_train['Embarked'] = dataset_train['Embarked'].replace(
    'S', 'Southampton')
sns.countplot(x='Embarked', data=dataset_train)

# View the Age distribution
sns.distplot(dataset_train.Age, bins=20)

# View the Fare distribution
sns.set_style('darkgrid')
sns.distplot(dataset_train.Fare)

# Check for independence between variables sex and survived
# We first create a contigency table
contingency_table = pd.crosstab(dataset_train['Sex'],
                                dataset_train['Survived'], margins=True)

# We then create an array with male / female values
femalecount = contingency_table.iloc[0].values
malecount = contingency_table.iloc[1].values
f_obs = np.array([femalecount, malecount])

# And then we input that array into the x^2 test function
scipy.stats.chi2_contingency(f_obs)[0:3]

# Based on our results we reject the null hypothesis. The variables
# are not independent.

# Subset the Dataset to exclude lines with no value for Age and Embarked
dataset_train_noNULLS = dataset_train[dataset_train['Age'].notnull(
) & dataset_train['Embarked'].notnull()]

# Create the independent and dependent variables
# Get the dummies to convert categorical variables to numerical
# Drop one dummy for each variable to avoid the dummy variable trap
cols = ['Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Embarked']
x_train = pd.get_dummies(dataset_train_noNULLS[cols])

x_train.drop('Pclass_3', axis=1, inplace=True)
x_train.drop('Embarked_Southampton', axis=1, inplace=True)
x_train.drop('Sex_male', axis=1, inplace=True)

y_train = dataset_train_noNULLS['Survived']

# Build a logreg and compute the feature importances
model = LogisticRegression(solver='liblinear')
model.fit(x_train, y_train)
model.coef_

# Try the RFE method
# Create the RFE model and select number of attributes
# We checked the appropriate number of attributes through the confusion matrix
rfe = RFE(model, 7).fit(x_train, y_train)

# summarize the selection of the attributes
print('Selected features: %s' % list(x_train.columns[rfe.support_]))

# Create a confusion matrix in the form of an array
y_pred = rfe.predict(x_train)
cnf_matrix = metrics.confusion_matrix(y_train, y_pred)
cnf_matrix

# Plot of predicted probabilities of survival vs actual survival
y_pred_prob = rfe.predict_proba(x_train)[:, 1]
plt.scatter(y_pred_prob, y_train, s=10)
plt.xlabel('Predicted Chance Of Survival')
plt.ylabel('Actual Survival')
plt.tight_layout()
plt.show()

print("Accuracy:", metrics.accuracy_score(y_train, y_pred))

#####################################################
# Test our model against dataset_test
# Fill out the missing values of age with the mean
# Fill out the missing values of fare with the median
mean_age = dataset_test.loc[:, 'Age'].mean()
print('The mean age in the test dataset is', round(mean_age, 1))
dataset_test['Age'] = dataset_test['Age'].fillna(round(mean_age, 1))

median_fare = dataset_test.loc[:, 'Fare'].median()
print('The median fare in the test dataset is', round(median_fare, 1))
dataset_test['Fare'] = dataset_test['Fare'].fillna(round(mean_age, 1))

# Format the test dataset as we did with the train set and get the dummies
dataset_test['Embarked'] = dataset_test['Embarked'].replace('C', 'Cherbourg')
dataset_test['Embarked'] = dataset_test['Embarked'].replace(
    'Q', 'Queenstown')
dataset_test['Embarked'] = dataset_test['Embarked'].replace(
    'S', 'Southampton')

dataset_test.Pclass = dataset_test.Pclass.astype('category')
x_test = pd.get_dummies(dataset_test[cols])

x_test.drop('Pclass_3', axis=1, inplace=True)
x_test.drop('Embarked_Southampton', axis=1, inplace=True)
x_test.drop('Sex_male', axis=1, inplace=True)

# Check for missing values
x_test.isnull().sum()

# Check that the shape of the train and test datasets is the same
x_test.shape
x_train.shape

# Make the predictions of survival in yes / no and in probabilities
y_pred_test = rfe.predict(x_test)

y_pred_prob_test = rfe.predict_proba(x_test)[:, 1]

dataset_actuals = pd.read_csv('data\\test survived actuals.csv')
y_actuals = dataset_actuals['Survived']
y_actuals = y_actuals.replace(
    0, 'No')
y_actuals = y_actuals.replace(
   1, 'Yes')


# Create a confusion matrix for the test-actual datasets
cnf_matrix = metrics.confusion_matrix(y_actuals, y_pred_test)
cnf_matrix

print("Accuracy:", round(metrics.accuracy_score(y_actuals, y_pred_test),3))
